## 38. 프로젝트 루트 확인

In [72]:
from pathlib import Path

project_root = Path.cwd()

print("현재 project 루트:", project_root)

if project_root.name == "notebooks":
    project_root = project_root.parent
    
data_dir = project_root / "data" / "raw"

print("프로젝트 루트:", project_root)
print("데이터 폴더:", data_dir)
print("데이터 폴더 존재:", data_dir.exists())

현재 project 루트: c:\dev\ai-data-analysis-repository\notebooks
프로젝트 루트: c:\dev\ai-data-analysis-repository
데이터 폴더: c:\dev\ai-data-analysis-repository\data\raw
데이터 폴더 존재: True


## 39. 필수 파일 확인

In [73]:
required_files = [
    "customers.csv",
    "products.csv",
    "orders.csv",
    "order_items.csv",
]

for file_name in required_files:
    file_path = data_dir / file_name
    print(file_name, file_path.exists(), file_path.stat().st_size if file_path.exists() else 0)

customers.csv True 5641
products.csv True 4101
orders.csv True 11696
order_items.csv True 15355


## 40. 데이터 불러오기

In [74]:
import pandas as pd
customers = pd.read_csv(data_dir / "customers.csv")
products = pd.read_csv(data_dir / "products.csv")
orders = pd.read_csv(data_dir / "orders.csv")
order_items = pd.read_csv(data_dir / "order_items.csv")

## 41. DataFrame 사전 만들기

In [75]:

datasets = {
    "customers": customers,
    "products": products,
    "orders": orders,
    "order_items": order_items,
}

## 42. 크기 요약

In [76]:
shape_summary = pd.DataFrame(

    [

        {
            "dataset": name,
            "rows": df.shape[0],
            "columns": df.shape[1],
        }

        for name, df in datasets.items()

    ]

)


print(shape_summary)

       dataset  rows  columns
0    customers   150        6
1     products   100        4
2       orders   301        5
3  order_items   765        5


## 43. 한 행의 의미 기록

## 데이터셋별 한 행의 의미

- customers: 고객 한 명
- products: 상품 한 개
- orders: 주문 한 건
- order_items: 주문에 포함된 상품 한 항목

 ## 44. 처음과 마지막 행 확인

In [77]:
for name, df in datasets.items():
    print(f"\n[{name}] 앞부분")
    display(df.head())
    print(f"[{name}] 뒷부분")
    display(df.tail())


[customers] 앞부분


,customer_id,name,gender,age,city,signup_date
0,1,김수민,F,19,광주,2024-07-17
1,2,김정호,F,32,대구,2025-11-30
2,3,이경수,F,61,성남,2024-07-10
3,4,조영호,F,55,울산,2026-05-11
4,5,이예원,F,19,부산,2024-10-11


[customers] 뒷부분


,customer_id,name,gender,age,city,signup_date
145,146,김숙자,M,61,성남,2026-01-20
146,147,이정남,M,19,부산,2025-03-11
147,148,오도현,M,29,고양,2026-07-13
148,149,김정자,M,20,부산,2024-11-16
149,150,조미영,M,40,대전,2026-01-01



[products] 앞부분


,product_id,product_name,category,price
0,1,전자기기 상품 001,전자기기,160000
1,2,도서 상품 002,도서,34000
2,3,전자기기 상품 003,전자기기,152000
3,4,생활용품 상품 004,생활용품,70000
4,5,식품 상품 005,식품,186000


[products] 뒷부분


,product_id,product_name,category,price
95,96,생활용품 상품 096,생활용품,112000
96,97,전자기기 상품 097,전자기기,154000
97,98,스포츠 상품 098,스포츠,10000
98,99,뷰티 상품 099,뷰티,200000
99,100,도서 상품 100,도서,102000



[orders] 앞부분


,order_id,customer_id,order_date,payment_method,order_status
0,1,123,2026-06-04,card,completed
1,2,77,2025-08-20,naver_pay,cancelled
2,3,138,2025-12-17,bank_transfer,cancelled
3,4,57,2026-02-27,kakao_pay,cancelled
4,5,125,2026-01-18,card,cancelled


[orders] 뒷부분


,order_id,customer_id,order_date,payment_method,order_status
296,297,22,2026-02-12,bank_transfer,completed
297,298,64,2026-01-22,naver_pay,cancelled
298,299,135,2025-09-05,bank_transfer,completed
299,300,96,2026-03-25,kakao_pay,cancelled
300,301,300,2026-09-08-06,kakao_pay,cancelled



[order_items] 앞부분


,order_item_id,order_id,product_id,quantity,unit_price
0,1,1,100,3,102000
1,2,1,87,5,25000
2,3,1,7,3,142000
3,4,1,9,3,193000
4,5,2,72,4,189000


[order_items] 뒷부분


,order_item_id,order_id,product_id,quantity,unit_price
760,761,298,40,4,174000
761,762,299,8,2,189000
762,763,299,12,4,175000
763,764,300,59,5,32000
764,765,499,200,3,128000


## 45. 컬럼명 확인

In [78]:

for name, df in datasets.items():
    print(name, df.columns.tolist())

customers ['customer_id', 'name', 'gender', 'age', 'city', 'signup_date']
products ['product_id', 'product_name', 'category', 'price']
orders ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status']
order_items ['order_item_id', 'order_id', 'product_id', 'quantity', 'unit_price']


## 46. 컬럼명 공백 점검

In [79]:
for name, df in datasets.items():

    suspicious = [
        col for col in df.columns
        if col != col.strip()
    ]

    print(name, suspicious)

customers []
products []
orders []
order_items []


## 47. info() 실행

In [80]:
for name, df in datasets.items():
    print(f"\n===== {name} =====")
    df.info()



===== customers =====
<class 'pandas.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   customer_id  150 non-null    int64
 1   name         150 non-null    str  
 2   gender       150 non-null    str  
 3   age          150 non-null    int64
 4   city         150 non-null    str  
 5   signup_date  150 non-null    str  
dtypes: int64(2), str(4)
memory usage: 7.2 KB

===== products =====
<class 'pandas.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   product_id    100 non-null    int64
 1   product_name  100 non-null    str  
 2   category      100 non-null    str  
 3   price         100 non-null    int64
dtypes: int64(2), str(2)
memory usage: 3.3 KB

===== orders =====
<class 'pandas.DataFrame'>
RangeIndex: 301 entries, 0 to 300
Data columns (total 5 columns):

## 48. dtype 요약표

In [81]:
dtype_rows = []

for name, df in datasets.items():
    for column, dtype in df.dtypes.items():
        dtype_rows.append(
            {
                "dataset": name,
                "column": column,
                "dtype": str(dtype),
            }
        )
dtype_summary = pd.DataFrame(dtype_rows)
dtype_summary

,dataset,column,dtype
0,customers,customer_id,int64
1,customers,name,str
2,customers,gender,str
3,customers,age,int64
4,customers,city,str
5,customers,signup_date,str
6,products,product_id,int64
7,products,product_name,str
8,products,category,str
9,products,price,int64


## 49. 날짜형 후보 확인

In [82]:

date_columns = {
    "customers": ["signup_date"],
    "orders": ["order_date"],
}

for dataset_name, columns in date_columns.items():
    df = datasets[dataset_name]
    for column in columns:
        converted = pd.to_datetime(
            df[column],
            errors="coerce"
        )

        print(
            dataset_name,
            column,
            "원본 dtype:", df[column].dtype,
            "변환 실패:", converted.isna().sum()
        )

customers signup_date 원본 dtype: str 변환 실패: 0
orders order_date 원본 dtype: str 변환 실패: 1


## 50. 결측치 개수

In [83]:
for name, df in datasets.items():
    print(f"\n[{name}]")
    display(df.isna().sum().to_frame("missing_count"))


[customers]


,missing_count
customer_id,0
name,0
gender,0
age,0
city,0
signup_date,0



[products]


,missing_count
product_id,0
product_name,0
category,0
price,0



[orders]


,missing_count
order_id,0
customer_id,0
order_date,0
payment_method,0
order_status,0



[order_items]


,missing_count
order_item_id,0
order_id,0
product_id,0
quantity,0
unit_price,0


## 51. 결측치 비율

In [84]:
missing_summaries = []

for name, df in datasets.items():
    summary = pd.DataFrame(
        {
            "missing_count": df.isna().sum(),
            "missing_rate_pct": (
                df.isna().mean() * 100
            ).round(2),
        }

    )

    summary.insert(0, "dataset", name)
    summary.insert(1, "column", summary.index)
    missing_summaries.append(summary.reset_index(drop=True))

missing_summary = pd.concat(
    missing_summaries,
    ignore_index=True
)

missing_summary

,dataset,column,missing_count,missing_rate_pct
0,customers,customer_id,0,0.0
1,customers,name,0,0.0
2,customers,gender,0,0.0
3,customers,age,0,0.0
4,customers,city,0,0.0
5,customers,signup_date,0,0.0
6,products,product_id,0,0.0
7,products,product_name,0,0.0
8,products,category,0,0.0
9,products,price,0,0.0


## 52. 전체 행 중복

In [85]:

for name, df in datasets.items():

    print(name, "전체 중복:", df.duplicated().sum())

customers 전체 중복: 0
products 전체 중복: 0
orders 전체 중복: 0
order_items 전체 중복: 0


## 53. 주요 ID 점검

In [86]:
primary_keys = {

    "customers": "customer_id",
    "products": "product_id",
    "orders": "order_id",
    "order_items": "order_item_id",

}

id_check_rows = []
for name, key in primary_keys.items():
    df = datasets[name]
    id_check_rows.append(
        {
            "dataset": name,
            "key": key,
            "missing": df[key].isna().sum(),
            "duplicates": df[key].duplicated().sum(),
            "unique": df[key].nunique(),
            "rows": len(df),
        }
    )

id_check_summary = pd.DataFrame(id_check_rows)
id_check_summary

,dataset,key,missing,duplicates,unique,rows
0,customers,customer_id,0,0,150,150
1,products,product_id,0,0,100,100
2,orders,order_id,0,0,301,301
3,order_items,order_item_id,0,0,765,765


## 54. 발견 내용을 기록합니다

## 데이터 품질 발견 기록

| 데이터셋        | 컬럼            | 문제 종류     | 건수 | 해석                        | 처리 여부                      |
| ----------- | ------------- | --------- | -: | ------------------------- | -------------------------- |
| customers   | signup_date   | 날짜 자료형    |  0 | 날짜 변환 실패는 없지만 현재 문자열 자료형임, datetime 변환 필요 | N/A |
| orders      | order_date    | 날짜 자료형    |  0 | 날짜 변환 실패는 없지만 현재 문자열 자료형임, datetime 변환 필요 | N/A |


## 55. 숫자형 기본 통계

In [87]:
print(customers[["age"]].describe())
print(products[["price"]].describe())
print(order_items[["quantity", "unit_price"]].describe())

              age
count  150.000000
mean    42.086667
std     15.613166
min     19.000000
25%     29.000000
50%     40.000000
75%     57.000000
max     69.000000
               price
count     100.000000
mean   110040.000000
std     56433.910574
min      5000.000000
25%     65750.000000
50%    112000.000000
75%    161000.000000
max    200000.000000
         quantity     unit_price
count  765.000000     765.000000
mean     3.053595  108586.928105
std      1.409951   56963.792455
min      1.000000    5000.000000
25%      2.000000   62000.000000
50%      3.000000  111000.000000
75%      4.000000  161000.000000
max      5.000000  200000.000000


## 56. 범주형 값 확인

In [88]:
category_columns = {

    "customers": ["gender", "city"],
    "products": ["category"],
    "orders": ["payment_method", "order_status"],

}

for dataset_name, columns in category_columns.items():
    df = datasets[dataset_name]
    for column in columns:
        print(f"\n[{dataset_name}.{column}]")
        display(
            df[column]
            .value_counts(dropna=False)
            .to_frame("count")
        )


[customers.gender]


,count
gender,
F,84
M,66



[customers.city]


,count
city,
성남,21
광주,17
부산,16
대구,15
서울,15
울산,14
인천,14
대전,14
수원,13



[products.category]


,count
category,
스포츠,19
전자기기,17
생활용품,16
뷰티,16
도서,14
패션,11
식품,7



[orders.payment_method]


,count
payment_method,
kakao_pay,80
naver_pay,77
bank_transfer,74
card,70



[orders.order_status]


,count
order_status,
completed,184
cancelled,65
refunded,52


## 57. 날짜 범위 확인

In [89]:

signup_dates = pd.to_datetime(

    customers["signup_date"],

    errors="coerce"

)

order_dates = pd.to_datetime(

    orders["order_date"],

    errors="coerce"

)

print("가입일 변환 실패:", signup_dates.isna().sum())

print("가입일 범위:", signup_dates.min(), "~", signup_dates.max())

print("주문일 변환 실패:", order_dates.isna().sum())

print("주문일 범위:", order_dates.min(), "~", order_dates.max())

가입일 변환 실패: 0
가입일 범위: 2023-08-11 00:00:00 ~ 2026-07-25 00:00:00
주문일 변환 실패: 1
주문일 범위: 2025-08-06 00:00:00 ~ 2026-08-05 00:00:00


## 58. 날짜 관계 점검 준비

주문일과 고객 가입일을 직접 비교하려면 두 데이터를 연결해야 합니다.

이번 장에서는 구조와 키가 연결 가능한지 확인하고, 실제 병합과 분석은 이후 장에서 본격적으로 진행합니다.


## 59. 참조하지 못한 고객 ID

In [90]:
unknown_customer_ids = sorted(
    set(orders["customer_id"].dropna())
    - set(customers["customer_id"].dropna())
)
print("존재하지 않는 customer_id:", unknown_customer_ids)

존재하지 않는 customer_id: [300]


## 60. 참조하지 못한 주문 ID

In [91]:
unknown_order_ids = sorted(
    set(order_items["order_id"].dropna())
    - set(orders["order_id"].dropna())
)
print("존재하지 않는 order_id:", unknown_order_ids)

존재하지 않는 order_id: [499]


## 61. 참조하지 못한 상품 ID

In [92]:
unknown_product_ids = sorted(
    set(order_items["product_id"].dropna())
    - set(products["product_id"].dropna())
)
print("존재하지 않는 product_id:", unknown_product_ids)

존재하지 않는 product_id: [200]


## 62. 반복 점검 함수

In [93]:
def inspect_dataframe(
    name: str,
    df: pd.DataFrame,
    primary_key: str | None = None,
) -> dict:
    result = {
        "dataset": name,
        "rows": df.shape[0],
        "columns": df.shape[1],
        "column_names": df.columns.tolist(),
        "dtypes": {column: str(dtype) for column, dtype in df.dtypes.items()},
        "missing_total": int(df.isna().sum().sum()),
        "duplicate_rows": int(df.duplicated().sum()),
    }
    if primary_key is not None:
        result["primary_key"] = primary_key
        result["key_missing"] = int(df[primary_key].isna().sum())
        result["key_duplicates"] = int(df[primary_key].duplicated().sum())
        result["key_unique"] = int(df[primary_key].nunique())
    return result

In [94]:
inspection_results = []
for name, df in datasets.items():
    inspection_results.append(
        inspect_dataframe(
            name,
            df,
            primary_keys[name],
        )
    )
inspection_results

[{'dataset': 'customers',
  'rows': 150,
  'columns': 6,
  'column_names': ['customer_id',
   'name',
   'gender',
   'age',
   'city',
   'signup_date'],
  'dtypes': {'customer_id': 'int64',
   'name': 'str',
   'gender': 'str',
   'age': 'int64',
   'city': 'str',
   'signup_date': 'str'},
  'missing_total': 0,
  'duplicate_rows': 0,
  'primary_key': 'customer_id',
  'key_missing': 0,
  'key_duplicates': 0,
  'key_unique': 150},
 {'dataset': 'products',
  'rows': 100,
  'columns': 4,
  'column_names': ['product_id', 'product_name', 'category', 'price'],
  'dtypes': {'product_id': 'int64',
   'product_name': 'str',
   'category': 'str',
   'price': 'int64'},
  'missing_total': 0,
  'duplicate_rows': 0,
  'primary_key': 'product_id',
  'key_missing': 0,
  'key_duplicates': 0,
  'key_unique': 100},
 {'dataset': 'orders',
  'rows': 301,
  'columns': 5,
  'column_names': ['order_id',
   'customer_id',
   'order_date',
   'payment_method',
   'order_status'],
  'dtypes': {'order_id': 'int6

## 63. LLM에 전달할 안전한 구조 요약

In [95]:
safe_schema_summary = {
    "datasets": inspection_results,
    "relationships": {
        "orders.customer_id": {
            "references": "customers.customer_id",
            "unknown_count": len(unknown_customer_ids),
        },
        "order_items.order_id": {
            "references": "orders.order_id",
            "unknown_count": len(unknown_order_ids),
        },
        "order_items.product_id": {
            "references": "products.product_id",
            "unknown_count": len(unknown_product_ids),
        },
    },
}
print(safe_schema_summary)

{'datasets': [{'dataset': 'customers', 'rows': 150, 'columns': 6, 'column_names': ['customer_id', 'name', 'gender', 'age', 'city', 'signup_date'], 'dtypes': {'customer_id': 'int64', 'name': 'str', 'gender': 'str', 'age': 'int64', 'city': 'str', 'signup_date': 'str'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'customer_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 150}, {'dataset': 'products', 'rows': 100, 'columns': 4, 'column_names': ['product_id', 'product_name', 'category', 'price'], 'dtypes': {'product_id': 'int64', 'product_name': 'str', 'category': 'str', 'price': 'int64'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'product_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 100}, {'dataset': 'orders', 'rows': 301, 'columns': 5, 'column_names': ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status'], 'dtypes': {'order_id': 'int64', 'customer_id': 'int64', 'order_date': 'str', 'payment_method': 'str', 'order_status'

## 64. LLM 구조 점검 프롬프트


나는 온라인 쇼핑몰 데이터 분석을 준비하고 있습니다.

아래 정보는 원본 행이 아니라 구조 점검 요약입니다.

[여기에 safe_schema_summary의 내용을 붙여 넣기]

다음을 검토해 주세요.

1. 각 데이터셋에서 한 행이 무엇을 의미할 가능성이 높은지 설명해 주세요.

2. 주요 ID와 외래 키를 구분해 주세요.

3. 즉시 삭제하면 안 되는 결측치·중복 사례를 알려 주세요.

4. 추가로 점검할 구조 문제를 우선순위대로 제안해 주세요.

5. 실제 데이터에 없는 컬럼이나 관계를 가정하지 마세요.

6. 각 제안에 대해 확인할 pandas 코드의 목적만 설명해 주세요.

개인정보나 원본 고객 행은 제공하지 않았습니다.

## 65. LLM 답변 검증표

| 검증 항목 | 확인 내용 |
|---|---|
|컬럼 존재|실제 컬럼명과 일치하는가?|
|한 행의 의미|데이터셋 역할과 일치하는가?|
|ID 해석|식별자를 수치 변수로 해석하지 않았는가?|
|중복 해석|반복 가능한 외래 키를 오류로 판단하지 않았는가?|
|결측 해석|업무 의미 확인 없이 삭제를 권하지 않는가?|
|관계|실제 연결 키와 일치하는가?|
|개인정보|추가 원본 행 제공을 요구하지 않는가?|

# 데이터 첫인상 보고서 요약본

> 원본 행이 아닌 구조 점검 요약을 기준으로 작성했습니다. 실제 값의 범위·분포가 필요한 항목은 추가 확인 대상으로 구분했습니다.

## 1. 데이터셋 및 컬럼·타입

### 데이터셋 구조

| 데이터셋          | 한 행의 의미          |   행 |  열 | 기본 키            | 외래 키                     |
| ------------- | ---------------- | --: | -: | --------------- | ------------------------ |
| `customers`   | 고객 1명            | 150 |  6 | `customer_id`   | 없음                       |
| `products`    | 상품 1개            | 100 |  4 | `product_id`    | 없음                       |
| `orders`      | 주문 1건            | 301 |  5 | `order_id`      | `customer_id`            |
| `order_items` | 주문에 포함된 상품 항목 1건 | 765 |  5 | `order_item_id` | `order_id`, `product_id` |

`orders`는 주문 단위이고, `order_items`는 주문에 포함된 상품 항목 단위입니다. 따라서 한 주문에 여러 상품이 포함되면 `order_items.order_id`가 여러 행에서 반복되는 것은 정상입니다.

### 컬럼과 타입

| 데이터셋          | 정수형 `int64`                                                         | 문자열형 `str`                                     | 주요 확인 사항                                    |
| ------------- | ------------------------------------------------------------------- | ---------------------------------------------- | ------------------------------------------- |
| `customers`   | `customer_id`, `age`                                                | `name`, `gender`, `city`, `signup_date`        | `customer_id`는 식별자, `signup_date`는 날짜 변환 필요 |
| `products`    | `product_id`, `price`                                               | `product_name`, `category`                     | `product_id`는 식별자, `price`는 범위 확인 필요        |
| `orders`      | `order_id`, `customer_id`                                           | `order_date`, `payment_method`, `order_status` | ID는 연결 키, `order_date`는 날짜 변환 필요            |
| `order_items` | `order_item_id`, `order_id`, `product_id`, `quantity`, `unit_price` | 없음                                             | ID와 수량·단가를 구분해서 해석해야 함                      |

ID 컬럼은 정수형이지만 계산용 숫자가 아니라 행을 식별하거나 파일을 연결하는 키입니다. 따라서 ID의 평균이나 합계에는 일반적인 분석 의미를 부여하지 않습니다.

---

## 2. 데이터 품질 점검 요약

| 점검 항목            | 현재 결과                              | 핵심 해석 및 추가 확인                                                                        |
| ---------------- | ---------------------------------- | ------------------------------------------------------------------------------------ |
| **결측치**          | 네 데이터셋 모두 0건                       | 일반 결측치는 없지만, 참조 대상에 없는 외래 키는 별도 문제로 확인해야 함                                           |
| **전체 중복**        | 네 데이터셋 모두 0건                       | 모든 컬럼값이 같은 완전 중복 행은 없음                                                               |
| **주요 ID 결측·중복**  | 네 기본 키 모두 결측 0건, 중복 0건             | 기본 키의 고유값 수가 행 수와 같아 각 행을 정상적으로 식별함                                                  |
| **숫자형 범위**       | 구조 요약만으로 확인 불가                     | `age`, `price`, `quantity`, `unit_price`의 음수·0·극단값을 확인해야 함                           |
| **범주형 값**        | 실제 고유값 미제공                         | `gender`, `city`, `category`, `payment_method`, `order_status`의 오탈자·공백·표기 혼재를 확인해야 함 |
| **날짜 범위와 변환 실패** | `signup_date`, `order_date`가 `str` | 날짜형 변환 실패 건수, 최소·최대 날짜, 가입일과 주문일의 순서를 확인해야 함                                         |
| **파일 간 참조 무결성**  | 세 관계에서 각각 불일치 1건                   | 관계별 불일치 수이며, 실제 고유 문제 행 수는 원본 데이터 확인이 필요함                                            |

### 주요 ID 상태

| 데이터셋          | 기본 키            | 결측 | 중복 | 고유값 수 | 판정 |
| ------------- | --------------- | -: | -: | ----: | -- |
| `customers`   | `customer_id`   |  0 |  0 |   150 | 정상 |
| `products`    | `product_id`    |  0 |  0 |   100 | 정상 |
| `orders`      | `order_id`      |  0 |  0 |   301 | 정상 |
| `order_items` | `order_item_id` |  0 |  0 |   765 | 정상 |

외래 키는 여러 행에서 반복될 수 있으므로 중복 여부보다 참조 대상이 실제로 존재하는지를 중심으로 점검해야 합니다.

### 숫자형·범주형·날짜 점검 대상

| 구분    | 컬럼                               | 확인 내용                     |
| ----- | -------------------------------- | ------------------------- |
| 숫자형   | `age`                            | 비현실적인 나이와 극단값             |
| 숫자형   | `price`, `unit_price`            | 음수·0·과도하게 큰 금액            |
| 숫자형   | `quantity`                       | 0 이하 또는 과도하게 큰 수량         |
| 범주형   | `gender`, `city`, `category`     | 고유값, 빈도, 공백 및 표기 일관성      |
| 범주형   | `payment_method`, `order_status` | 동일 의미의 표현 분리 여부           |
| 날짜    | `signup_date`, `order_date`      | 변환 실패, 날짜 형식 혼재, 최소·최대 날짜 |
| 날짜 관계 | `signup_date`, `order_date`      | 주문일이 가입일보다 앞선 사례          |

0원 가격이나 단가는 무료·증정 상품일 수 있으므로 확인 없이 오류로 판단하지 않습니다. 날짜 순서가 맞지 않는 경우도 데이터 이관 등 업무상 이유가 있을 수 있어 원인 확인이 필요합니다.

### 파일 간 참조 무결성

| 외래 키                     | 참조 대상                   | 관계별 불일치 |
| ------------------------ | ----------------------- | ------: |
| `orders.customer_id`     | `customers.customer_id` |      1건 |
| `order_items.order_id`   | `orders.order_id`       |      1건 |
| `order_items.product_id` | `products.product_id`   |      1건 |

세 외래 키 관계에서 각각 참조 불일치가 1건씩 확인되었습니다. 다만 하나의 `order_items` 행에서 `order_id`와 `product_id`가 동시에 일치하지 않을 수도 있으므로, 실제 문제 행이 반드시 3개라고 단정할 수는 없습니다.

---

## 3. 발견한 문제

| 구분     | 발견 내용                                         | 현재 판단                             |
| ------ | --------------------------------------------- | --------------------------------- |
| 확인된 문제 | `orders.customer_id` 참조 불일치 1건                | 원인 확인 필요                          |
| 확인된 문제 | `order_items.order_id` 참조 불일치 1건              | 주문 정보 누락 또는 ID 오류 가능성 확인 필요       |
| 확인된 문제 | `order_items.product_id` 참조 불일치 1건            | 상품 기준정보 누락 또는 ID 오류 가능성 확인 필요     |
| 추가 점검  | 날짜 컬럼이 문자열 타입                                 | 날짜 변환 및 실패값 확인 필요                 |
| 추가 점검  | 숫자형 실제 범위 미확인                                 | 음수·0·극단값 확인 필요                    |
| 추가 점검  | 범주형 실제 값 미확인                                  | 오탈자·공백·표기 혼재 확인 필요                |
| 추가 점검  | `products.price`와 `order_items.unit_price` 관계 | 현재 가격과 주문 당시 단가일 수 있어 차이 원인 확인 필요 |
| 추가 점검  | 주문 상세가 없는 주문                                  | 반대 방향의 연결 여부 확인 필요                |
| 추가 점검  | 동일 주문·상품 조합 반복                                | 정상 분할 행인지 중복 적재인지 확인 필요           |

기본 키 중복, 전체 행 중복, 일반 결측치는 발견되지 않았습니다. 현재 가장 명확한 문제는 세 외래 키 관계의 참조 불일치입니다.

---

## 4. 즉시 처리하지 않은 이유

| 대상                       | 즉시 삭제·수정하지 않는 이유                                         |
| ------------------------ | -------------------------------------------------------- |
| 외래 키 불일치 행               | 추출 시점 차이, 기준정보 누락, 과거 기록, ID 입력 오류, 테스트 데이터 등 여러 원인이 가능함 |
| 반복되는 외래 키                | 한 고객의 여러 주문, 한 주문의 여러 상품처럼 정상적인 관계에서 발생할 수 있음            |
| 동일 고객 이름·상품명             | 서로 다른 고객이나 상품이 같은 이름을 사용할 수 있음                           |
| 0원 가격·단가                 | 무료 상품, 증정품 또는 프로모션일 가능성이 있음                              |
| `price`와 `unit_price` 차이 | 현재 상품 가격과 주문 당시 판매 단가가 다를 수 있음                           |
| 동일 주문·상품 조합              | 같은 상품을 별도 항목으로 기록했을 수도 있어 자동 중복 삭제가 위험함                  |
| 날짜 순서 이상                 | 데이터 이관이나 가입 정보 기준 차이 등 업무상 원인이 있을 수 있음                   |

따라서 이상 후보는 바로 제거하지 않고 별도 검토 대상으로 분리한 후 원인에 따라 수정·보완·제외 여부를 결정하는 것이 안전합니다.

---

## 5. 다음 전처리 단계에서 결정할 내용

| 우선순위 | 결정할 내용        | 주요 작업                                              |
| ---: | ------------- | -------------------------------------------------- |
|    1 | 참조 무결성 처리 기준  | 관계별 불일치 행과 실제 고유 문제 행 수 확인                         |
|    2 | 불일치 원인별 처리 방식 | ID 수정, 기준정보 보완, 분석 제외 또는 별도 보관 결정                  |
|    3 | 날짜 처리 기준      | 날짜형 변환, 실패값 관리, 분석 기간 설정                           |
|    4 | 숫자형 유효 범위     | `age`, `price`, `quantity`, `unit_price`의 허용 범위 결정 |
|    5 | 범주형 표준화       | 공백 제거, 오탈자 확인, 동일 의미 값 통합                          |
|    6 | 가격 사용 기준      | 상품 기준 가격과 주문 당시 단가의 역할 구분                          |
|    7 | 중복 후보 판단 기준   | 동일 주문·상품 조합을 정상 반복과 적재 오류로 구분                      |
|    8 | 병합 및 집계 기준    | 주문 수는 고유 `order_id`로 계산하고 조인 후 중복 집계 방지            |

## 종합 의견

현재 데이터는 **기본 키, 일반 결측치, 전체 중복 측면에서는 전반적으로 양호**합니다. 가장 먼저 확인할 사항은 세 외래 키 관계에서 각각 발견된 참조 불일치 1건이며, 실제 고유 문제 행 수와 발생 원인을 확인한 후 처리해야 합니다.

그다음 날짜 변환, 숫자형 범위, 범주형 표기, 가격 컬럼의 의미를 점검하고, 마지막으로 테이블 병합과 집계 기준을 확정하는 순서가 적절합니다.